src/state_point.cpp

```diff
@@ -14,13 +15,294 @@
+    case RUN_MODE_FIXEDSOURCE:
+      write_dataset(file_id, "run_mode", "fixed source");
+      break;
+    case RUN_MODE_EIGENVALUE:
+      write_dataset(file_id, "run_mode", "eigenvalue");
+      break;
+    }
+    write_attribute(file_id, "photon_transport", settings::photon_transport);
+    write_dataset(file_id, "n_particles", settings::n_particles);
+    write_dataset(file_id, "n_batches", settings::n_batches);
+
+    // Write out current batch number
+    write_dataset(file_id, "current_batch", simulation::current_batch);
+
+    // Indicate whether source bank is stored in statepoint
+    write_attribute(file_id, "source_present", write_source_);
+
+    // Write out information for eigenvalue run
+    if (settings::run_mode == RUN_MODE_EIGENVALUE) write_eigenvalue_hdf5(file_id);
+
+    // Write tally-related data
+    hid_t tallies_group = create_group(file_id, "tallies");
+    meshes_to_hdf5(tallies_group);
+    statepoint_write_f(file_id);
+    close_group(tallies_group);
+  }
+
+  // Check for the no-tally-reduction method
+  if (!settings::reduce_tallies) {
+    // If using the no-tally-reduction method, we need to collect tally
+    // results before writing them to the state point file.
+
+    write_tally_results_nr(file_id);
+
+  } else if (mpi::master) {
+
+    // Write number of global realizations
+    write_dataset(file_id, "n_realizations", n_realizations);
+
+    // Write out the runtime metrics.
+    using namespace simulation;
+    hid_t runtime_group = create_group(file_id, "runtime");
+    write_dataset(runtime_group, "total initialization",  time_initialize.elapsed());
+    write_dataset(runtime_group, "reading cross sections", time_read_xs.elapsed());
+    write_dataset(runtime_group, "simulation", time_inactive.elapsed()
+      + time_active.elapsed());
+    write_dataset(runtime_group, "transport", time_transport.elapsed());
+    if (settings::run_mode == RUN_MODE_EIGENVALUE) {
+      write_dataset(runtime_group, "inactive batches", time_inactive.elapsed());
+    }
+    write_dataset(runtime_group, "active batches", time_active.elapsed());
+    if (settings::run_mode == RUN_MODE_EIGENVALUE) {
+      write_dataset(runtime_group, "synchronizing fission bank", time_bank.elapsed());
+      write_dataset(runtime_group, "sampling source sites", time_bank_sample.elapsed());
+      write_dataset(runtime_group, "SEND-RECV source sites", time_bank_sendrecv.elapsed());
+    }
+    write_dataset(runtime_group, "accumulating tallies", time_tallies.elapsed());
+    write_dataset(runtime_group, "total", time_total.elapsed());
+    close_group(runtime_group);
+
+    file_close(file_id);
+  }
+
+#ifdef PHDF5
+  bool parallel = true;
+#else
+  bool parallel = false;
+#endif
+
+  // Write the source bank if desired
+  if (write_source_) {
+    if (mpi::master || parallel) file_id = file_open(filename_, 'a', true);
+    write_source_bank(file_id);
+    if (mpi::master || parallel) file_close(file_id);
+  }
+
+  return 0;
```

Comment:
Since this function is part of the C API, the default is always to return an error code. In this case, it doesn't look like there's any other return value, but that doesn't mean there won't be in the future, so we should keep the return type the same.

Reply:
Ah, right on.